### PR TITLE
Add offline mode

### PR DESCRIPTION
**Pull request by wp0pw**
Add offline mode

Added one new option
 - `--data-local` to provide a path to a file a dir containing mmr logs

Renamed `--url` to `--data-url`

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -34,6 +34,12 @@ func NewApp() *cli.App {
 			&cli.BoolFlag{
 				Name: "envauth", Usage: "set to enable authorization from the environment (not all commands support this)",
 			},
+			&cli.StringFlag{
+				Name: "logdir", Usage: "set to a path to directory containing log files belonging to single mmr to use that data instead of remote logs",
+			},
+			&cli.StringFlag{
+				Name: "logfile", Usage: "set to a path to file or '-' for stdin to use specific mmr log instead of remote logs",
```
Comment (robinbryce) [inline app.go:41]: Its minor, but is there scope for confusion --loglevel and --logdir ?

File: cfgmassif.go [hunk 2]
```diff
@@ -24,14 +24,29 @@ func cfgMassifReader(cmd *CmdCtx, cCtx *cli.Context) error {
 		}
 	}
 
-	if cmd.reader == nil {
-		if err = cfgReader(cmd, cCtx); err != nil {
+	logdir := cCtx.String("logdir")
```
Comment (robinbryce) [inline cfgmassif.go:27]: Why not use an argv positional argument and just test if it is a file or a directory ? Its more natural on the command line I think

File: cfgmassif.go [hunk 3]
```diff
@@ -24,14 +24,29 @@ func cfgMassifReader(cmd *CmdCtx, cCtx *cli.Context) error {
...
+		if err != nil {
+			return err
+		}
+		cmd.massifReader = mr
+	} else {
+		reader, err := cfgReader(cmd, cCtx)
+		if err != nil {
 			return err
 		}
+		mr := massifs.NewMassifReader(logger.Sugar, reader, massifs.WithoutGetRootSupport())
```
Comment (robinbryce) [inline cfgmassif.go:47]: It is probably a bug that we had WithoutGetRootSupport set, that means the peak stack is not mapped in the context and so attempts to refer to the peaks needed from the earlier massifs will fail

File: app.go (hunk 1 above)
Comment (wp0pw) [reply app.go:41]: yeh, definitely - happy to change the name of the option to something more sensible

File: cfgmassif.go (hunk 2 above)
Comment (wp0pw) [reply cfgmassif.go:27]: so I think we already use positional args for event.json and what not - then the order would start to matter too and it's all potentially bit messy - guess we should probably agree on a sensible approach though that will work for all the commands

File: localmassifreader.go [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+	Open(string) (io.ReadCloser, error)
+}
+
+type DirLister interface {
+	// ListFiles returns list of absolute paths
+	// to files (not subdirectories) in a directory
+	ListFiles(string) ([]string, error)
+}
+
+type LocalMassifReader struct {
```
Comment (honourfish) [inline localmassifreader.go:28]: awesome :) i love that we have a local massif reader now, is it worth putting it in the massifs module?

File: localmassifreader.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+	return mc, nil
+}
+
+// NewLocalMassifReader creates MassifReader that reads from
+// local files on disc - it mostly ignores tenant identity
+// as we assume all the logs on the disc are for the tenant one is
+// interested in - but it's still valid to pass tenant ID here
+func NewLocalMassifReader(
+	log logger.Logger, opener Opener, dirLister DirLister, logdir string, logfile string, opts ...massifs.MassifReaderOption,
```
Comment (honourfish) [inline localmassifreader.go:106]: is it worth making logdir and logfile options? or having a New() function for each?

There maybe a time when you want one or the other, not both?

File: localmassifreader.go (hunk 4 above)
Comment (wp0pw) [reply localmassifreader.go:28]: I was thinking about that but the interface is kind of rough as we still expose the azure blob stuff quite a bit - I think once we have time to maybe review the interfaces on both ends we can refactor and pull this one into the massifs module

File: localmassifreader.go (hunk 5 above)
Comment (wp0pw) [reply localmassifreader.go]: I just changed the interface here a bit - so now it's only local data and a bool for dir/not dir

Comment (wp0pw) [reply localmassifreader.go]: added options now for directory

File: cfgmassif.go [hunk 6]
```diff
@@ -24,19 +26,57 @@ func cfgMassifReader(cmd *CmdCtx, cCtx *cli.Context) error {
...
 
+	localLog := cCtx.String("data-local")
+	remoteLog := cCtx.String("data-url")
+
+	if localLog != "" && remoteLog != "" {
+		return fmt.Errorf("can't use data-local and data-url at the same time")
+	}
+
+	if localLog == "" && remoteLog == "" {
+		// if we had no url and no local data supplied we try STDIN
```
Comment (henry739) [inline cfgmassif.go:42]: Might become evident later in review, but how does this interact with commands that take events on stdin?

File: cfgmassif.go [hunk 7]
```diff
@@ -24,19 +26,57 @@ func cfgMassifReader(cmd *CmdCtx, cCtx *cli.Context) error {
...
+		absLocal, err := filepath.Abs(localLog)
+		if err != nil {
+			return err
+		}
+		fi, err := os.Stat(absLocal)
+		if err != nil {
+			return err
+		}
+		opts := []Option{}
+		if fi.IsDir() {
```
Comment (henry739) [inline cfgmassif.go:61]: Potentially, we can fail here if its not a directory

Comment (wp0pw) [reply cfgmassif.go:61]: we also accept a single file

**Review by henry739 — APPROVED** [review]

**Pull request merged**